src/elliptic/curves/secp256_k1.rs

```diff
@@ -54,415 +74,251 @@ const BASE_POINT2_Y: [u8; 32] = [
-        Secp256k1Point {
-            purpose: "random_point",
-            ge: pk.get_element(),
-        }
+
+impl ops::Deref for PK {
+    type Target = PublicKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl ops::DerefMut for PK {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
     }
 }
 
-impl Zeroize for Secp256k1Scalar {
+impl Zeroize for SK {
     fn zeroize(&mut self) {
-        unsafe { ptr::write_volatile(self, FE::zero()) };
+        let sk = self.0.as_mut_ptr();
+        let sk_bytes = unsafe { std::slice::from_raw_parts_mut(sk, 32) };
+        sk_bytes.zeroize()
+    }
+}
+
+impl Zeroize for PK {
+    fn zeroize(&mut self) {
+        let zeroed = unsafe { secp256k1::ffi::PublicKey::new() };
+        unsafe { ptr::write_volatile(self.0.as_mut_ptr(), zeroed) };
         atomic::fence(atomic::Ordering::SeqCst);
         atomic::compiler_fence(atomic::Ordering::SeqCst);
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum Secp256k1 {}
+
+impl Curve for Secp256k1 {
+    type Point = GE;
+    type Scalar = FE;
+
+    fn curve_name() -> &'static str {
+        "secp256k1"
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Secp256k1Scalar {
+    purpose: &'static str,
+    /// Zeroizing<SK> wraps SK and zeroize it on drop
+    ///
+    /// `fe` might be None — special case for scalar being zero
+    fe: zeroize::Zeroizing<Option<SK>>,
+}
+#[derive(Clone, Debug, Copy)]
+pub struct Secp256k1Point {
+    purpose: &'static str,
+    ge: Option<PK>,
+}
+
+type GE = Secp256k1Point;
+type FE = Secp256k1Scalar;
+
 impl ECScalar for Secp256k1Scalar {
-    type SecretKey = SK;
+    type Underlying = Option<SK>;
 
-    fn new_random() -> Secp256k1Scalar {
+    fn random() -> Secp256k1Scalar {
+        let sk = SK(SecretKey::new(&mut thread_rng()));
         Secp256k1Scalar {
             purpose: "random",
-            fe: SecretKey::new(&mut thread_rng()),
+            fe: Zeroizing::new(Some(sk)),
         }
     }
 
     fn zero() -> Secp256k1Scalar {
-        let zero_arr = [0u8; 32];
-        let zero = unsafe { std::mem::transmute::<[u8; 32], SecretKey>(zero_arr) };
         Secp256k1Scalar {
             purpose: "zero",
-            fe: zero,
+            fe: Zeroizing::new(None),
         }
     }
 
-    fn get_element(&self) -> SK {
-        self.fe
+    fn is_zero(&self) -> bool {
+        self.fe.is_none()
     }
 
-    fn set_element(&mut self, element: SK) {
-        self.fe = element
-    }
+    fn from_bigint(n: &BigInt) -> Secp256k1Scalar {
+        if n.is_zero() {
+            return Self::zero();
+        }
 
-    fn from(n: &BigInt) -> Secp256k1Scalar {
-        let curve_order = FE::q();
-        let n_reduced = BigInt::mod_add(n, &BigInt::from(0), &curve_order);
-        let mut v = BigInt::to_bytes(&n_reduced);
+        let curve_order = Self::curve_order();
+        let n_reduced = n.modulus(curve_order);
```

Comment:
What do you think about returning an error if `n` is bigger than the curve order instead of reducing it?

Reply:
All the code we have now relies on `from_bigint` reducing `n`. I think if we just change behaviour, something will get broken without notifying us! But it might make sense to add method `try_from_bigint`, what do you think?